spec/controllers/catalog_controller_spec.rb

```diff
@@ -32,5 +32,47 @@
         expect(controller).to have_search_parameters
       end
     end
+
+    context "when not logged in" do
```

Comment:
So I think there are three options:

1) Test the controllers as integration specs. This means setting up the required objects + users appropriately, but not rendering views. They're reasonably fast, and will break if some sort of interface issue arises, which I think is good.

2) Build full integration specs, either as request specs or Capybara specs. They're much slower, but hit the full stack, so you know if it's broken.

3) Treat controller specs as units. Mock out things like `Ability#can?`, loading of resources, etc, so you can get a clear look at the interfaces you're depending on, and build out from there.

It's been my experience that #1 supplemented by a very small amount of #2 is the best way to test -RAILS- controllers, largely because it has this huge assumption that all these things are going to be interconnected at that point. I've never successfully found a way to make those interfaces explicit, since there's no parameter passing allowed. At least, I haven't without doing something like "response objects", where you extract each action off into some sort of functional construct, and that's always hard to follow and tends to hurt adoption from other Rails developers.

I'd be totally open to a PR exploring those ideas though, or being part of a pair to start trying to make those ideas explicit, because I agree that doing it this way hurts a lot of the point of TDD as an architecture exploration technique.